### PR TITLE
UGC Market

### PR DESCRIPTION
**Pull request by Zintixx**
Implements UGC Market.
A couple of things to note is that the way the packets are set up, the game normally is supposed to have UGC items and Meret Market items in one table/class. 
I think this is messy and they should be segregated. So that's what's done here.

Currently not working is:
- Listing UGC Mounts or blocks (because you can't make them yet)
- searching functionality
- "Most Popular" sorting. Currently just sorts by most purchases

### CODE REVIEW

File: GameDataParser/Parsers/MeretMarketCategoryParser.cs [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+            foreach (XmlNode node in document.DocumentElement.ChildNodes)
+            {
+                if (node.Name != "category")
+                {
+                    continue;
+                }
+
+                MeretMarketSection section = (MeretMarketSection) int.Parse(node.Attributes["id"].Value);
+
+
```
Comment (AlanMorel) [inline GameDataParser/Parsers/MeretMarketCategoryParser.cs:34]: Extra line

**Review by AlanMorel — APPROVED** [review]

File: MapleServer2/Database/Classes/DatabaseItem.cs [hunk 2]
```diff
@@ -218,6 +220,7 @@ private Item ReadItem(dynamic data)
             PairedCharacterId = data.paired_character_id,
             PairedCharacterName = data.paired_character_name,
             PetSkinBadgeId = data.pet_skin_badge_id,
+            RecommendJobs = ItemMetadataStorage.GetRecommendJobs(data.id),
```
Comment (AngeloTadeucci) [inline MapleServer2/Database/Classes/DatabaseItem.cs:223]: Isn't this already happening in `Item.SetMetadataValues()`?

**Review by AngeloTadeucci — CHANGES_REQUESTED** [review]

File: MapleServer2/Types/UGCMarketItem.cs [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    public UGCMarketListingStatus Status;
+    public long CreationTimestamp;
+    public long ListingExpirationTimestamp;
+    public long PromotionExpirationTimestamp;
+    public long SellerAccountId;
+    public long SellerCharacterId;
+    public string SellerCharacterName;
+    public string Description;
+    public int SalesCount;
+    public List<string> Tags = new List<string>();
```
Comment (AngeloTadeucci) [inline MapleServer2/Types/UGCMarketItem.cs:22]:
```suggestion
    public List<string> Tags = new();
```

File: MapleServer2/PacketHandlers/Game/MeretMarketHandler.cs [hunk 4]
```diff
@@ -55,27 +90,274 @@ public override void Handle(GameSession session, PacketReader packet)
...
+        {
+            return;
+        }
+
+        if (salePrice < item.UGC.SalePrice || salePrice > long.Parse(ConstantsMetadataStorage.GetConstant("UGCShopSellMaxPrice")))
+        {
+            return;
+        }
+
+        long totalFee = GetListingFee(session.Player.CharacterId, promote);
```
Comment (AngeloTadeucci) [inline MapleServer2/PacketHandlers/Game/MeretMarketHandler.cs:142]: not being used

File: MapleServer2/PacketHandlers/Game/MeretMarketHandler.cs [hunk 5]
```diff
@@ -55,27 +90,274 @@ public override void Handle(GameSession session, PacketReader packet)
...
+        string description = packet.ReadUnicodeString();
+        long listingFee = packet.ReadLong();
+
+        UGCMarketItem item = GameServer.UGCMarketManager.FindItemById(ugcMarketItemId);
+        if (item is null || item.SellerCharacterId != session.Player.CharacterId || item.ListingExpirationTimestamp < TimeInfo.Now())
+        {
+            return;
+        }
+
+        long totalFee = GetListingFee(session.Player.CharacterId, promote);
```
Comment (AngeloTadeucci) [inline MapleServer2/PacketHandlers/Game/MeretMarketHandler.cs:220]: not being used

File: MapleServer2/PacketHandlers/Game/MeretMarketHandler.cs [hunk 6]
```diff
@@ -55,27 +90,274 @@ public override void Handle(GameSession session, PacketReader packet)
...
-        byte[] unk = packet.ReadBytes(12);
+        long ugcItemId = packet.ReadLong();
+        if (ugcItemId != 0)
+        {
+            PurchaseUGCItem(session, ugcItemId);
+            return;
+        }
+
+        PurchasePremiumItem(session, packet, marketItemId);
+        return;
```
Comment (AngeloTadeucci) [inline MapleServer2/PacketHandlers/Game/MeretMarketHandler.cs:328]:
```suggestion
```

File: MapleServer2/PacketHandlers/Game/MeretMarketHandler.cs [hunk 7]
```diff
@@ -55,27 +90,274 @@ public override void Handle(GameSession session, PacketReader packet)
...
+        }
+
+        if (!HandleMarketItemPay(session, marketItem.Price, MeretMarketCurrencyType.Meret))
+        {
+            return;
+        }
+
+        marketItem.SalesCount++;
+        DatabaseManager.UGCMarketItems.Update(marketItem);
+        UGCMarketSale sale = new(marketItem.Price, marketItem.Item.UGC.Name, marketItem.SellerCharacterId);
```
Comment (AngeloTadeucci) [inline MapleServer2/PacketHandlers/Game/MeretMarketHandler.cs:346]: not being used

File: MapleServer2/Managers/UGCMarketManager.cs [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+    }
+
+    public void RemoveListing(UGCMarketItem item)
+    {
+        Items.Remove(item.Id);
+    }
+
+    public List<UGCMarketItem> GetItemsByCharacterId(long characterId)
+    {
+        return new(Items.Values.Where(b => b.SellerCharacterId == characterId).ToList());
```
Comment (AngeloTadeucci) [inline MapleServer2/Managers/UGCMarketManager.cs:45]:
```suggestion
        return Items.Values.Where(b => b.SellerCharacterId == characterId).ToList();
```

File: MapleServer2/Managers/UGCMarketManager.cs [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+    }
+
+    public List<UGCMarketItem> GetPromoItems()
+    {
+        List<UGCMarketItem> items = Items.Values.Where(x => x.PromotionExpirationTimestamp > TimeInfo.Now() && x.Status == UGCMarketListingStatus.Active).ToList();
+        Random random = RandomProvider.Get();
+        items = items.OrderBy(x => random.Next()).ToList();
+        return items.Take(12).ToList(); // 12 being the max the shop can display
+    }
```
Comment (AngeloTadeucci) [inline MapleServer2/Managers/UGCMarketManager.cs:54]: I feel this is easier to read. You can be the judge
```suggestion
    public List<UGCMarketItem> GetPromoItems()
    {
        return Items.Values.Where(x => x.PromotionExpirationTimestamp > TimeInfo.Now() && x.Status == UGCMarketListingStatus.Active)
            .OrderBy(_ => RandomProvider.Get().Next()).Take(12).ToList(); // 12 being the max the shop can display
    }
```

File: MapleServer2/Managers/UGCMarketManager.cs [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+
+    public List<UGCMarketItem> GetNewestItems()
+    {
+        List<UGCMarketItem> items = new(Items.Values.OrderBy(x => x.ListingExpirationTimestamp).ToList());
+        return items.Where(x => x.Status == UGCMarketListingStatus.Active).Take(6).ToList();
+    }
```
Comment (AngeloTadeucci) [inline MapleServer2/Managers/UGCMarketManager.cs:60]: There is no need to use a `new List()` then also use `.toList()`
```suggestion
    public List<UGCMarketItem> GetNewestItems()
    {
        return Items.Values.OrderBy(x => x.ListingExpirationTimestamp).Where(x => x.Status == UGCMarketListingStatus.Active).Take(6).ToList();
    }
```

File: MapleServer2/Managers/UGCMarketManager.cs [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+            {
+                Gender gender;
+                if (genderFlag.HasFlag(GenderFlag.Male))
+                {
+                    gender = Gender.Male;
+                }
+                else
+                {
+                    gender = Gender.Female;
+                }
```
Comment (AngeloTadeucci) [inline MapleServer2/Managers/UGCMarketManager.cs:95]:
```suggestion
                Gender gender = genderFlag.HasFlag(GenderFlag.Male) ? Gender.Male : Gender.Female;
```

File: MapleServer2/Managers/UGCMarketManager.cs [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
+    }
+
+    public void RemoveSale(UGCMarketSale sale)
+    {
+        Sales.Remove(sale.Id);
+    }
+
+    public List<UGCMarketSale> GetSalesByCharacterId(long characterId)
+    {
+        return new(Sales.Values.Where(b => b.SellerCharacterId == characterId).ToList());
```
Comment (AngeloTadeucci) [inline MapleServer2/Managers/UGCMarketManager.cs:142]:
```suggestion
        return Sales.Values.Where(b => b.SellerCharacterId == characterId).ToList();
```

**Review by AngeloTadeucci — CHANGES_REQUESTED** [review]

File: MapleServer2/Database/Classes/DatabaseItem.cs (hunk 2 above)
Comment (Zintixx) [reply MapleServer2/Database/Classes/DatabaseItem.cs:223]: Nope, `DataseItem` uses the empty constructer for `Item`

File: MapleServer2/PacketHandlers/Game/MeretMarketHandler.cs (hunk 7 above)
Comment (Zintixx) [reply MapleServer2/PacketHandlers/Game/MeretMarketHandler.cs:346]: What do you mean? it makes a new `UGCMarketSale` and within the constructer it adds it to the `UGCMarketManager`

Comment (AngeloTadeucci) [reply MapleServer2/PacketHandlers/Game/MeretMarketHandler.cs:346]: I meant the variable is not being used

**Review by AngeloTadeucci — APPROVED** [review]

**Review by darkvergus — APPROVED** [review]

**Pull request merged**